internal/config/timeperiod.go

```diff
@@ -1,100 +1,51 @@
-	stmt = r.db.BuildSelectStmt(entryPtr, entryPtr)
-	r.logger.Debugf("Executing query %q", stmt)
-
-	var entries []*timeperiod.Entry
-	if err := tx.SelectContext(ctx, &entries, stmt); err != nil {
-		r.logger.Errorln(err)
-		return err
-	}
-
-	for _, entry := range entries {
-		p := timePeriodsById[entry.TimePeriodID]
-		if p == nil {
-			r.logger.Warnw("ignoring entry for unknown timeperiod_id",
-				zap.Int64("timeperiod_entry_id", entry.ID),
-				zap.Int64("timeperiod_id", entry.TimePeriodID))
-			continue
-		}
-
-		if p.Name == "" {
-			p.Name = fmt.Sprintf("Time Period #%d", entry.TimePeriodID)
-		}
-
-		err := entry.Init()
-		if err != nil {
-			r.logger.Warnw("ignoring time period entry",
-				zap.Object("entry", entry),
-				zap.Error(err))
-			continue
-		}
-
-		p.Entries = append(p.Entries, entry)
-
-		r.logger.Debugw("loaded time period entry",
-			zap.Object("timeperiod", p),
-			zap.Object("entry", entry))
-	}
-
-	for _, p := range timePeriodsById {
-		if p.Name == "" {
-			p.Name = fmt.Sprintf("Time Period #%d (empty)", p.ID)
-		}
-	}
-
-	if r.TimePeriods != nil {
-		// mark no longer existing time periods for deletion
-		for id := range r.TimePeriods {
-			if _, ok := timePeriodsById[id]; !ok {
-				timePeriodsById[id] = nil
-			}
-		}
-	}
-
-	r.pending.TimePeriods = timePeriodsById
-
-	return nil
-}
-
 func (r *RuntimeConfig) applyPendingTimePeriods() {
-	if r.TimePeriods == nil {
-		r.TimePeriods = make(map[int64]*timeperiod.TimePeriod)
-	}
+	incrementalApplyPending(
+		r,
+		&r.TimePeriods, &r.configChange.TimePeriods,
+		func(newElement *timeperiod.TimePeriod) error {
+			if newElement.Name == "" {
+				newElement.Name = fmt.Sprintf("Time Period #%d", newElement.ID)
+			}
+			return nil
+		},
+		func(element, update *timeperiod.TimePeriod) error {
+			if update.Name != "" {
+				element.Name = update.Name
+			}
+			return nil
+		},
+		nil)
+
+	incrementalApplyPending(
+		r,
+		&r.timePeriodEntries, &r.configChange.timePeriodEntries,
+		func(newElement *timeperiod.Entry) error {
+			period, ok := r.TimePeriods[newElement.TimePeriodID]
+			if !ok {
+				return fmt.Errorf("time period entry refers unknown time period %d", newElement.TimePeriodID)
+			}
 
-	for id, pendingTimePeriod := range r.pending.TimePeriods {
-		if pendingTimePeriod == nil {
-			delete(r.TimePeriods, id)
-		} else if currentTimePeriod := r.TimePeriods[id]; currentTimePeriod != nil {
-			*currentTimePeriod = *pendingTimePeriod
-		} else {
-			r.TimePeriods[id] = pendingTimePeriod
-		}
-	}
+			period.Entries = append(period.Entries, newElement)
+			return nil
+		},
+		nil,
+		func(delElement *timeperiod.Entry) error {
+			period, ok := r.TimePeriods[delElement.TimePeriodID]
+			if !ok {
+				return fmt.Errorf("time period entry refers unknown time period %d", delElement.TimePeriodID)
+			}
 
-	r.pending.TimePeriods = nil
+			period.Entries = slices.DeleteFunc(period.Entries, func(entry *timeperiod.Entry) bool {
+				return entry.ID == delElement.ID
+			})
```

Comment:
I tried not to change code directly outside of this change's scope too much. On the one hand, it would simplify those apply methods where I have to toy around with the `slices` package, but on the other hand, the diff would be bigger and reach more in other packages. Thus, I would prefer to keep it this way.